PHPLIB-1309 Add addSubscriber/removeSubscriber to Client class to ease configuration (#1195)
mongodb/mongo-php-library: docs/reference/method/MongoDBClient-addSubscriber.txt

================================
MongoDB\\Client::addSubscriber()
================================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

Definition
----------

.. phpmethod:: MongoDB\\Client::addSubscriber()

    Registers a monitoring event subscriber with this Client. The subscriber
    will be notified of all events for this Client.

   .. code-block:: php

      function addSubscriber(MongoDB\Driver\Monitoring\Subscriber $subscriber): void

   This method has the following parameters:

   .. include:: /includes/apiargs/MongoDBClient-method-addSubscriber-param.rst

   .. note::

         If $subscriber is already registered with this Client, this function
         is a no-op. If $subscriber is also registered globally, it will still
         only be notified once of each event for this Client.

Errors/Exceptions
-----------------

> .. include:: /includes/extracts/error-invalidargumentexception.rst

Author: This exception is duplicated in the extension [doc](https://www.php.net/manual/en/mongodb-driver-manager.addsubscriber.php):
> Throws [MongoDB\Driver\Exception\InvalidArgumentException](https://www.php.net/manual/en/class.mongodb-driver-exception-invalidargumentexception.php) if subscriber is a [MongoDB\Driver\Monitoring\LogSubscriber](https://www.php.net/manual/en/class.mongodb-driver-monitoring-logsubscriber.php), since loggers can only be registered globally.

How should I render it here?

================================
MongoDB\\Client::addSubscriber()
================================

> .. default-domain:: mongodb

Reviewer: ```suggestion
.. versionadded:: 1.18

.. default-domain:: mongodb
```

See other method pages for prior art.

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

Definition
----------

.. phpmethod:: MongoDB\\Client::addSubscriber()

    Registers a monitoring event subscriber with this Client. The subscriber
    will be notified of all events for this Client.

   .. code-block:: php

      function addSubscriber(MongoDB\Driver\Monitoring\Subscriber $subscriber): void

   This method has the following parameters:

   .. include:: /includes/apiargs/MongoDBClient-method-addSubscriber-param.rst

   .. note::

>          If $subscriber is already registered with this Client, this function
>          is a no-op. If $subscriber is also registered globally, it will still

Reviewer: ```suggestion
         If ``$subscriber`` is already registered with this Client, this function
         is a no-op. If ``$subscriber`` is also registered globally, it will still
```

Other pages use monospace formatting around argument names.

         only be notified once of each event for this Client.

Errors/Exceptions
-----------------

.. include:: /includes/extracts/error-invalidargumentexception.rst

Example
-------

Create a :phpclass:`MongoDB\\Driver\\Monitoring\\CommandSubscriber` that

> logs all events:

Author: I created this example as the example on [Application Performance Monitoring (APM)](https://www.php.net/manual/en/mongodb.tutorial.apm.php#mongodb.tutorial.apm) seemed complicated to me

Author: I update the example to use the client method. Here is the ticket for the tutorial: https://jira.mongodb.org/browse/PHPLIB-1314

            {
                fwrite($this->stream, sprintf(
                    'Succeeded command #%d "%s" in %d microseconds: %s%s',
                    $event->getRequestId(),
                    $event->getCommandName(),
                    $event->getDurationMicros(),
                    json_encode($event->getReply()),
                    PHP_EOL,
                ));
            }

            public function commandFailed(CommandFailedEvent $event): void
            {
                fwrite($this->stream, sprintf(
                    'Failed command #%d "%s" in %d microseconds: %s%s',
                    $event->getRequestId(),
                    $event->getCommandName(),
                    $event->getDurationMicros(),
                    $event->getError()->getMessage(),
                    PHP_EOL,
                ));
            }
        }

The subscriber can then be registered with a Client:

.. code-block:: php

    <?php

        $client = new MongoDB\Client();
        $subscriber = new LogCommandSubscriber(STDERR);

        $client->addSubscriber($subscriber);

        $client->test->users->insertOne(['username' => 'alice']);

> The above code will log the following to err output:

Reviewer: ```suggestion
The above code will write the following to stderr output:
```

I'd refrain from using "log" here, since command logging will eventually be implemented down the road. That will be different from APM in that the driver will generate the log message on its own.

.. code-block:: text

    Started command #1 "insert": { "insert" : "users", "ordered" : true, "$db" : "test", "lsid" : { "id" : { "$binary" : { "base64" : "dKTBhZD7Qvi0vUhvR58mCA==", "subType" : "04" } } }, "documents" : [ { "username" : "alice", "_id" : { "$oid" : "655d1fca12e81018340a4fc2" } } ] }
    Succeeded command #1 "insert" in 3349 microseconds: {"n":1,"ok":1}

See Also
--------

- :phpmethod:`MongoDB\\Client::removeSubscriber()`

> - :php:`Application Performance Monitoring (APM)

Reviewer: No objection to make these lines longer for readability. There is prior art for that, such as `MongoDBClient-getWriteConcern.txt`.

  <manual/en/mongodb.tutorial.apm>`
- :php:`MongoDB\\Driver\\Manager::addSubscriber()
  <manual/en/mongodb-driver-manager.addsubscriber>`
- :php:`MongoDB\Driver\Monitoring\Subscriber

>   <manual/en/class.mongodb-driver-monitoring-subscriber>`

Reviewer: I think you can drop this since it will already be linked from the API docs above with my earlier suggestion. Likewise for the remove method.

Author: Yes, and `CommandSubscriber` is more interesting.
